fix: Improve cluster connection pool logic when disconnecting

martinslota opened the pull request:
This PR is a port of https://github.com/redis/ioredis/pull/1864 and what follows below is a verbatim copy of that PR's description. A reproduce for the bug described here that uses `valkey_server` and `iovalkey` instead of `redis_server` and `ioredis`, respectively, can be found https://github.com/martinslota/redis-connection-race-bug.

# Motivation and Background

This is an attempt to fix errors occurring when a `connect()` call is made shortly after a `disconnect()`, which is something that [the Bull library does when pausing a queue](https://github.com/OptimalBits/bull/blob/60fa88f08637f0325639988a3f054880a04ce402/lib/queue.js#L869-L871).

Here's a relatively minimal way to reproduce an error:

```js
import IORedis from "ioredis";

const cluster = new IORedis.Cluster([{ host: "localhost", port: 6380 }]);

await cluster.set("foo", "bar");

const endPromise = new Promise((resolve) => cluster.once("end", resolve));
await cluster.quit();
cluster.disconnect();
await endPromise;

cluster.connect();
console.log(await cluster.get("foo"));
cluster.disconnect();
```

Running that script in a loop using 

```sh
#!/bin/bash

set -euo pipefail

while true
do
    DEBUG=ioredis:cluster node cluster-error.mjs
done
```

against the `main` branch of `ioredis` quickly results in this output:

```
/Code/ioredis/built/cluster/index.js:124
                    reject(new redis_errors_1.RedisError("Connection is aborted"));
                           ^

RedisError: Connection is aborted
    at /Code/ioredis/built/cluster/index.js:124:28

Node.js v20.11.0
```

My debugging led me to believe that the existing node cleanup logic in the `ConnectionPool` class leads to race conditions: upon [`disconnect()`](https://github.com/redis/ioredis/blob/ec42c82ceab1957db00c5175dfe37348f1856a93/lib/cluster/index.ts#L283-L304), the [this.connectionPool.reset() call](https://github.com/redis/ioredis/blob/ec42c82ceab1957db00c5175dfe37348f1856a93/lib/cluster/index.ts#L302) will [remove nodes from the pool](https://github.com/redis/ioredis/blob/ec42c82ceab1957db00c5175dfe37348f1856a93/lib/cluster/ConnectionPool.ts#L129) without cleaning up [the event listener](https://github.com/redis/ioredis/blob/ec42c82ceab1957db00c5175dfe37348f1856a93/lib/cluster/ConnectionPool.ts#L90-L96) which may then subsequently issue more than one `drain` event. Depending on timing, one of the extra `drain` events may fire after `connect()` and [change the status to `close`](https://github.com/redis/ioredis/blob/ec42c82ceab1957db00c5175dfe37348f1856a93/lib/cluster/index.ts#L153), interfering with the connection attempt and leading to the error above.

# Changes

* Keep track of node listeners in the `ConnectionPool` class and remove them from the nodes whenever they are removed from the pool.
* Issue `-node` / `drain` regardless of whether nodes disconnected or were removed through a `reset()` call.
* Within `reset()`, add nodes before removing old ones to avoid unwanted `drain` events.
* Fix one of the listeners by using an arrow function to make `this` point to the connection pool instance.
* Try to fix the script for running cluster tests and attempt to enable them on CI. If this doesn't work out or isn't useful, I'm happy to revert the changes.
* Add a test around this issue. The error thrown in the test on `main` is seemingly different from the error shown above but it still seems related to the disconnection logic and still gets fixed by the changes in this PR.

mcollina reviewed (approved):
lgtm with a green CI

martinslota commented:
Now that I saw the results from CI, I also figured out how to run the unit tests locally. 😅 

The tests caught a bug that I then fixed in https://github.com/valkey-io/iovalkey/pull/5/commits/788361c043695aac31d5d4bd0e34ae9d6a57593e and https://github.com/valkey-io/iovalkey/pull/5/commits/0deaebaecdb6daa65295674ea19d14b4ad9fdbd9. 😬 

Furthermore, some of the existing functional tests were assuming that responding with `'OK'` from the mock server would work fine as a response to `cluster SLOTS`, i.e. that the `Cluster` client would reach a `ready` state despite the response being invalid. With the changes made in this PR, the tests were failing because part of `connectionPool.reset([])` is now the `drain` event which leads to the `disconnecting` state instead, after which the client attempts to reconnect over and over again, never reaching the `ready` state.

In my view, this change in behaviour is desirable since it prevents node clients which are no longer tracked in the connection pool from emitting additional events that could mess with the state of the cluster client (which could easily have moved on, e.g. attempting to reconnect). So for the time being, I adjusted the tests by adding valid slots tables to the mock server in https://github.com/valkey-io/iovalkey/pull/5/commits/d5d85e95b9d97f915778ce0bd0d533f41a5bc6aa.

Finally, one of the tests was assuming that when a node disappears from the cluster, the node removal (the `-node` event) would only be emitted *after* the `refreshSlotsCache()` method was finished executing. Once again, this assumption is not valid with the changes in this PR because the event is emitted already during the execution of `refreshSlotsCache()`, and again as part of `connectionPool.reset([])`. For similar reasons as above, I adjusted the test in https://github.com/valkey-io/iovalkey/pull/5/commits/d3f83c52cf9ef88d3ca8271356fd7f574777dad6 so that it starts listening for a node removal event already before it calls `refreshSlotsCache()`.

Does this approach sound kinda reasonable?

mcollina reviewed (approved):
Yes.

Still LGTM

martinslota commented:
As far as I can tell, this should be ready to get merged. I cannot merge it myself.

mcollina commented:
Will ship a release asap.
